Spatial_Engine/Query/DominantVector.cs

```diff
@@ -0,0 +1,154 @@
+            int biggestOrthogonalLengthIndex = orthogonalLengths.IndexOf(orthogonalLengths.Max());
+            
+            Vector orthogonalDominantVector = orthogonalVectors[biggestOrthogonalLengthIndex].First().Normalise();
+            if (dominantVector.IsEqual(orthogonalDominantVector))
+                return dominantVector;
+            
+            //check if length tolerance passes
+            if (orthogonalLengths.Max() < groupedLengths.Max() * orthogonalLengthTolerance)
+                return dominantVector;
+
+            return orthogonalDominantVector;
+        }
+
+        /******************************************/
+        /****            IElement2D            ****/
+        /******************************************/
+        
+        [Description("Gets the the dominant vector (orientation) of an Element2D based on its outter lines lengths.")]
+        [Input("element2D", "Element2D to evaluate.")]
+        [Input("orthogonalPriority", "Optional, if true gives priority to curves that are on the orthogonal axis.")]
+        [Input("orthogonalTolerance", "Optional, when orthogonalPriority is true it will only return orthogonal vector when its length is higher this number multiplied by the actual dominant vector lengths. For example if the dominant vector is 10 in length but the orthogonal is only 5 in length, then this number should be 0.5.")]
+        [Input("angleTolerance", "Optional, angle in radians that vectors will be considered similar. Default value in radians is approximately 5 degrees.")]
+        [Output("dominantVector", "The the dominant vector of an Element1D.")]
+        public static BH.oM.Geometry.Vector DominantVector(IElement2D element2D, bool orthogonalPriority = true, double orthogonalLengthTolerance = 0.5, double angleTolerance = 0.087)
+        {
+            IElement1D outline = BH.Engine.Geometry.Create.PolyCurve(element2D.IOutlineElements1D().Select(x =>x as ICurve));
+            return DominantVector(outline, orthogonalPriority, orthogonalLengthTolerance, angleTolerance);
+        }
+        
+        /******************************************/
+        /****              Private             ****/
+        /******************************************/
+
+        [Description("Groups vectors by direction whilst allowing for an angle discrepancy tolerance.")]
+        [Input("vectors", "Vectors to evaluate.")]
+        [Input("angleTolerance", "The angle in radians to compare vectors with each other for tolerance when grouping.")]
+        [Output("GroupSimilarVectorsWithTolerance", "The grouped vectors.")]
+        private static List<List<Vector>> GroupSimilarVectorsWithTolerance(List<Vector> vectors, double angleTolerance)
+        {
+            List<List<Vector>> result = new List<List<Vector>>();
+            List<Vector> orderByLength = vectors.OrderBy(x => x.Length()).ToList();
+
+            for (int i = 0; i < orderByLength.Count; i++)
```

Comment:
This `for` loop looks a bit overcooked to me:
- `i` will never be different than 0 (you increment and decrement by 1 at each iteration)
- you could start `j` from 1 to avoid checking whether vector[i] is equal to vector [j]

Not crucial in this case, but can be a good mental excercise to clean this up a bit.